### DISCUSSION TITLE
unsupported MediaType: "application/vnd.docker.distribution.manifest.v1+prettyjws"

### DISCUSSION THREAD

**Discussion by martin-g** [concourse/concourse#6439] [Help & Support] [answered]
Hi,

I have a web node (v. 6.7.3, running on x86_64) and a worker node (from https://github.com/neumayer/concourse-arm-worker)

The web node is started as:
```
#!/usr/bin/env fish

set -x CONCOURSE_POSTGRES_HOST localhost
set -x CONCOURSE_POSTGRES_USER concourse_user
set -x CONCOURSE_POSTGRES_PASSWORD concourse_pass
set -x CONCOURSE_POSTGRES_DATABASE concourse
set -x CONCOURSE_EXTERNAL_URL http://x64:8080
set -x CONCOURSE_X_FRAME_OPTIONS deny
set -x CONCOURSE_ADD_LOCAL_USER user:pass
set -x CONCOURSE_MAIN_TEAM_LOCAL_USER user
set -x CONCOURSE_WORKER_BAGGAGECLAIM_DRIVER overlay
set -x CONCOURSE_SESSION_SIGNING_KEY ./keys/session_signing_key
set -x CONCOURSE_TSA_HOST_KEY ./keys/tsa_host_key
set -x CONCOURSE_TSA_AUTHORIZED_KEYS ./keys/authorized_worker_keys
set -x CONCOURSE_BIND_IP 0.0.0.0
set -x CONCOURSE_PEER_URL http://0.0.0.0:8080
set -x CONCOURSE_AUTH_DURATION 60m
set -x CONCOURSE_CLUSTER_NAME concourse-on-arm64 

./bin/concourse web
```

Note: In Fish shell `set -x` is like `export` in Bash.

The worker node is started as a Docker container:
```
#!/usr/bin/fish

docker run  \
  -it \
  --rm \
  -p 7777:7777 \
  -p 7788:7788 \
  -p 8888:8888 \
  -v /home/ubuntu/devel/concourse:/concourse \
  -e CONCOURSE_NAME=arm64-worker \
  -e CONCOURSE_WORK_DIR=/concourse/worker \
  -e CONCOURSE_TSA_PUBLIC_KEY=/concourse/keys/tsa_host_key.pub \
  -e CONCOURSE_TSA_WORKER_PRIVATE_KEY=/concourse/keys/worker_key \
  -v /home/ubuntu/devel/concourse/keys:/concourse-keys \
  --privileged \
  -e CONCOURSE_TSA_HOST=192.168.0.206:2222 \
  -e CONCOURSE_RUNTIME=containerd \
  concourse-arm-worker worker
```

```
$  fly -t tutorial workers -d
name          containers  platform  tags  team  state    version  age  garden address   baggageclaim url        active tasks  resource types
arm64-worker  0           linux     none  none  running  2.2      4s   127.0.0.1:46637  http://127.0.0.1:41947  0             registry-image
```

When I try to run the following task:
```

---
platform: linux

image_resource:
  #type: docker-image     # concourse-arm-worker does not support docker-image
  type: registry-image
  source: {repository: busybox}

run:
  path: echo
  args: [hello world]
```

I get this error:

```
$ fly -t tutorial execute -c ~/git/concourse-tutorial/tutorials/basic/task-hello-world/task_hello_world.yml
executing build 36 at http://127.0.0.1:8080/builds/36 
initializing
fetching busybox@sha256:2780635f864cc66c7a5c74aca8047970b95cb91b6d5c135964d984ffe07a2024
ERRO[0002] download failed: get image: unsupported MediaType: "application/vnd.docker.distribution.manifest.v1+prettyjws", see https://github.com/google/go-containerregistry/issues/377 
fetching the image did not produce a volume
errored
```

Ideas what could be the problem ?

**Comment by taylorsilva** [comment, marked as answer]
tldr: the image being pulled has a v1 manifest and only v2 is supported.

Error message mentions this issue: https://github.com/google/go-containerregistry/issues/377 which led me to find this issue in the `registry-image` https://github.com/concourse/registry-image-resource/issues/19

I confirmed this locally by doing:
```
docker pull busybox@sha256:2780635f864cc66c7a5c74aca8047970b95cb91b6d5c135964d984ffe07a2024
```
I got this warning from docker:
```
Image docker.io/library/busybox@sha256:2780635f864cc66c7a5c74aca8047970b95cb91b6d5c135964d984ffe07a2024 
uses outdated schema1 manifest format.
Please upgrade to a schema2 image for better future compatibility.
More information at https://docs.docker.com/registry/spec/deprecated-schema-v1/
```

It seems the arm image for busybox is using an old manifest. Not sure why the are but they probably should upgrade their image. That seems to be the general sentiment from https://github.com/concourse/registry-image-resource/issues/19

**Reply by martin-g** [reply]
The actual problem here is that without specifying source's `tag` ConcourseCI will download random (?!) version of the image. As a Docker user I expected that it will download the `latest` tag. Once I specified `latest` I got the following error:
```
running echo hello world
Backend error: Exit status: 500, message: {"Type":"","Message":"proc start: OCI runtime exec failed: exec failed: container_linux.go:349: starting container process caused \"exec format error\": unknown","Handle":"","ProcessID":"","Binary":""}
```
which is saying that I am using a x86_64 Docker image and the aarch64 Docker daemon cannot run it.

Updating my task YAML to:
```

---
platform: linux

image_resource:
  type: registry-image
  source:
    repository: arm64v8/busybox
    tag: latest

run:
  path: echo
  args: [hello world]
```
finally ran it successfully:
```
executing build 42 at http://127.0.0.1:8080/builds/42 
initializing
fetching arm64v8/busybox@sha256:bf920ca7f146b802e1c9a8aab1fba3a3fe601c56b075ecef90834c13b90bb5bb
c2cf70554e62 [======================================] 797.8KiB/797.8KiB
selected worker: arm64-worker
running echo hello world
hello world
succeeded
```

Thank you for your help, @taylorsilva !

**Reply by taylorsilva** [reply]
I wouldn't say a random image was fetched. It probably looked at the `latest` tag first and saw that the arch was wrong, and then proceeded to look at all digests available for the image and fetched the first digest that matched your workers arch.

By specifying the `latest` tag you forced it to download whatever image `latest` points to, ignoring the fact that the arch is wrong.

Glad you fixed it by finding an arm busybox image!

**Reply by martin-g** [reply]
I've just retried it with a freshly started worker (i.e. an empty worker/volumes/) and commented out `tag: latest` in the YAML.

```
fly -t tutorial e -c ~/git/concourse-tutorial/tutorials/basic/task-hello-world/task_hello_world.yml
executing build 55 at http://127.0.0.1:8080/builds/55 
initializing
fetching arm64v8/busybox@sha256:a65c1c37a94fb28407eb4c1716e80ab7ecd231d8641c1b380aa851cd5f4f9b84
593072949377 [======================================] 690.0KiB/690.0KiB
selected worker: arm64-worker
running echo hello world
hello world
succeeded
```

But `arm64v8/busybox:latest` resolves to https://hub.docker.com/layers/arm64v8/busybox/latest/images/sha256-bf920ca7f146b802e1c9a8aab1fba3a3fe601c56b075ecef90834c13b90bb5bb?context=explore

I.e. 
- downloaded: `a65c1c37a94fb28407eb4c1716e80ab7ecd231d8641c1b380aa851cd5f4f9b84`
- latest:            `bf920ca7f146b802e1c9a8aab1fba3a3fe601c56b075ecef90834c13b90bb5bb`

Re-enabling `tag: latest` in the YAML leads to:
```
$ fly -t tutorial e -c ~/git/concourse-tutorial/tutorials/basic/task-hello-world/task_hello_world.yml
executing build 56 at http://127.0.0.1:8080/builds/56 
initializing
fetching arm64v8/busybox@sha256:bf920ca7f146b802e1c9a8aab1fba3a3fe601c56b075ecef90834c13b90bb5bb
c2cf70554e62 [======================================] 797.8KiB/797.8KiB
selected worker: arm64-worker
running echo hello world
hello world
succeeded
```

So, I am not sure whether it is random tag or not, but definitely it is not downloading the `latest` !

**Reply by aoldershaw** [reply]
I think the issue is that the concourse-arm-worker is using the latest registry-image-resource (based on master of that repo), which isn't actually compatible with 6.x.

There was a bug in the image fetching code in 6.x and earlier that doesn't affect the official registry-image resource bundled with concourse, but is incompatible with the new registry-image resource - I believe it tries to use the *oldest* digest tagged with a semver tag.

This behaviour will be fixed in the next release (7.0)

**Reply by martin-g** [reply]
Thank you for explaining the issue, @aoldershaw !
Then this suggested fix by @neumayer should also help: https://github.com/neumayer/concourse-arm-worker/pull/3#issuecomment-763616575 ?

**Reply by aoldershaw** [reply]
It's important to checkout a specific tag, but the v1.0.0 tag will cause problems until the concourse version is bumped to 7.0.0 (for the reasons mentioned above)

I'd suggest checking out 0.14.0 instead - that's the same latest version that we bundle with any currently released version of concourse

**Reply by neumayer** [reply]
Yeah I was just about to say at least it's not unspecified after that change of mine anymore :-)

The only viable long term strategy is to start making a pull request that updates the official build pipelines to support the arm version (so all the versioning stuff is taking care of).

Maybe we can find the time at some point to make that contribution.

**Reply by martin-g** [reply]
> The only viable long term strategy is to start making a pull request that updates the official build pipelines to support the arm version (so all the versioning stuff is taking care of).

I do have some spare time to help with this!
I just need some pointers how to do it!

**Reply by aoldershaw** [reply]
Yeah, it'd definitely be a maintenance nightmare trying to match our pipelines! We intend to add ARM support eventually, but not sure when we'd be able to get to it ourselves - so if you'd be willing to contribute, that'd be awesome!

All of our CI configuration is in the https://github.com/concourse/ci repo. Specifically, the pipeline we'd need to update is the release pipeline (https://github.com/concourse/ci/blob/master/pipelines/release.yml).

Someone previously proposed a PR for adding support for ARM binaries (https://github.com/concourse/ci/pull/386/files), but it doesn't look like it works in its current state - it builds the binary, but it doesn't actually do anything with it, nor does it build any resource types (esp. registry-image). Also looks like there's [an issue with pipeline syntax](https://github.com/concourse/ci/pull/386/files#diff-e6fc109a45234e08636c0723683a184cadc9ceb4baa2d86b097dd7e937d8ab60R573-R575).

So, some things we'll need to do are:
1. Build the concourse binary for ARM and upload it to GCS in the `build-rc` job
    * The existing PR does this, but the `params` syntax is a bit off - it should use an object instead of a list
    * Also, we'll need to build it for alpine and ubuntu - not just alpine, as is done in that PR
2. Build the resource types (or at least the registry-image resource) for ARM (this happens outside the release pipeline)
    * I'm not actually sure how to build images for ARM. Do you need to be running on an ARM machine? Does the Dockerfile need to change at all?
    * In any event, we'll need to mess around with the resources pipeline template (https://github.com/concourse/ci/blob/master/pipelines/resources/template.yml) (it uses [ytt](https://github.com/vmware-tanzu/carvel-ytt) for generating the pipeline YAML for each resource type), and possibly the Dockerfiles in the resource type repos (e.g. https://github.com/concourse/registry-image-resource/blob/master/dockerfiles/alpine/Dockerfile) to build ARM versions of the resource types based off of both alpine and ubuntu
3. In the `resource-types-images` job, we'll need to bundle up the ARM resource types somehow
    * This could mean creating two new images (`resource-types-alpine-arm-image`, `resource-types-ubuntu-arm-image`) if we choose to build all of the resource types with ARM
    * If we just build the registry-image resource with ARM, maybe it makes sense to just add it to the existing images - we just use these images as an intermediate packaging stage if I'm not mistaken
    * Note: the eventual goal is to *not* bundle any resource types (except for registry-image) along with concourse - https://github.com/concourse/rfcs/pull/30. So, maybe just building registry-image for ARM is good enough for now
4. Build Concourse's dependencies for ARM
    * Currently, we just bundle supported releases of our dependencies based on their github releases
    * The dependencies are Guardian (the current default runtime), and containerd/runc/cni plugins (for the currently experimental containerd runtime, which will eventually become the default)
    * Until these dependencies start releasing ARM variants in their github releases, I'm not sure the best approach to take here - perhaps we'd need to checkout the associated tag and build them ourselves?
5. In the `build-rc-image` job, we need to use the ARM concourse binaries and the ARM resource types (possibly just registry-image) to build a alpine+linux variants for an ARM concourse image
6. The `create-draft-release` job will need to `get` the built concourse binaries from GCS (same as from step 1), as will the `publish-binaries` job. The `prep-release-assets` task in `publish-binaries` will also need to copy the ARM binaries to the `concourse-linux` output

There's still a lot I'm not sure about, but I hope that helps if you're interested in contributing!

**Comment by robinhuiser** [comment]
I've created a [repository](https://github.com/robinhuiser/concourse-arm64) to build `arm64` images.

This repository helps you build both the web and worker `arm64` components for Concourse CI - prebuilt Docker images can be found on Docker Hub [rdclda/concourse](https://hub.docker.com/repository/docker/rdclda/concourse).

## Bundled resources

| Concourse | [git](https://github.com/concourse/git-resource) | [github-release](https://github.com/concourse/github-release-resource) | [registry-image](https://github.com/concourse/registry-image-resource) | [semver](https://github.com/concourse/semver-resource) | [time](https://github.com/concourse/time-resource) | [mock](https://github.com/concourse/mock-resource) | [s3](https://github.com/concourse/s3-resource) | [slack-alert](https://github.com/arbourd/concourse-slack-alert-resource) |
|--- |--- |--- |--- |--- |--- |--- |--- |--- |
| v7.1.0 | v1.12.0 | v1.5.2 | v1.2.0 | v1.3.0 | v1.5.0 | v0.11.1 | v1.1.1 | v0.15.0 |
| v7.2.0 | v1.12.1 | v1.5.2 | v1.2.1 | v1.3.0 | v1.6.0 | v0.11.1 | v1.1.1 | v0.15.0 |
| v7.3.2 | v1.14.0 | v1.6.1 | v1.3.0 | v1.3.1 | v1.6.0 | v0.11.2 | v1.1.1 | v0.15.0 |
| v7.4.0 | v1.14.0 | v1.6.4 | v1.4.0 | v1.3.4 | v1.6.1 | v0.12.2 | v1.1.2 | v0.15.0 |
| v7.5.0 | v1.14.4 | v1.6.4 | v1.4.1 | v1.3.4 | v1.6.2 | v0.12.3 | v1.1.3 | v0.15.0 |
